### PR TITLE
StatObject with Args & Response objects.

### PR DESCRIPTION
**Pull request by BigUstad**
Example changed accordingly.

### CODE REVIEW

File: Minio/Helper/OperationsHelper.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+using System.Collections.Immutable;
+
+namespace Minio
+{
+    public class OperationsUtil
+    {
+        internal static bool IsSupportedHeader(string hdr, IEqualityComparer<string> comparer = null)
+        {
+            comparer = comparer ?? StringComparer.OrdinalIgnoreCase;
+            var supportedHeaders = ImmutableArray.Create<string>(new string[]{ "cache-control", "content-encoding", "content-type", "x-amz-acl", "content-disposition" });
```
Comment (poornas) [inline Minio/Helper/OperationsHelper.cs:28]: there seems to be 2 other definitions for supported headers - why does this need to be available as a public class. Can you unify them? This is internal to sdk

**Review by poornas — CHANGES_REQUESTED** [review]

File: Minio/Helper/OperationsHelper.cs (hunk 1 above)
Comment (BigUstad) [reply Minio/Helper/OperationsHelper.cs:28]: Yes. I was eventually going to replace supportedHeaders. I will do that in this PR itself by using this function.

File: Minio.Examples/Cases/StatObjectQuery.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
+/*
+ * MinIO .NET Library for Amazon S3 Compatible Cloud Storage, (C) 2017 MinIO, Inc.
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio.Examples/Cases/StatObjectQuery.cs:2]: Done.

File: Minio/DataModel/ObjectStat.cs [hunk 3]
```diff
@@ -46,9 +50,21 @@ public ObjectStat(string objectName, long size, DateTime lastModified, string et
...
         public override string ToString()
         {
+            if (!string.IsNullOrEmpty(this.VersionId))
+            {
+                string versionInfo = $"Version ID({this.VersionId})";
+                if (this.DeleteMarker)
+                {
+                    versionInfo = $"Version ID({this.VersionId}, deleted)";
+                }
+                return $"{this.ObjectName} :  Size({this.Size}) LastModified({this.LastModified}) ETag({this.ETag}) Content-Type({this.ContentType})";
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/DataModel/ObjectStat.cs:66]: Changed. DeleteMarker is also applied.

File: Minio.Examples/Cases/StatObjectQuery.cs [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            {
+                Console.WriteLine("Running example for API: StatObjectAsync [with ObjectQuery]");
+                StatObjectArgs args = new StatObjectArgs()
+                                                .WithBucket(bucketName)
+                                                .WithObject(bucketObject)
+                                                .WithVersionId(versionID)
+                                                .WithMatchETag(matchEtag)
+                                                .WithNotMatchETag(notMatchEtag)
+                                                .WithModifiedSince(modifiedSince)
+                                                .WithUnModifiedSince(unModifiedSince);
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio.Examples/Cases/StatObjectQuery.cs:66]: I've added the checks both in the example & also in the Validate() function. I've kept the .With conditions part as-is to keep the example code small.

File: Minio.Examples/Cases/StatObjectQuery.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        // Get stats on a object
+        public async static Task Run(MinioClient minio, 
+                                     string bucketName = "my-bucket-name",
+                                     string bucketObject = "my-object-name",
+                                     string versionID = null,
+                                     string matchEtag = null,
+                                     string notMatchEtag = null,
+                                     DateTime modifiedSince = default(DateTime),
+                                     DateTime unModifiedSince = default(DateTime))
+        {
```
Comment (poornas) [inline Minio.Examples/Cases/StatObjectQuery.cs:44]:
```suggestion
        public async static Task Run(MinioClient minio, 
                                     string bucketName = "my-bucket-name",
                                     string bucketObject = "my-object-name",
                                     string versionID = null,
                                     string matchEtag = null,
                                     //string notMatchEtag = null,
                                     DateTime modifiedSince = default(DateTime),
                                     //DateTime unModifiedSince = default(DateTime)
                                     )
 ```

File: Minio.Examples/Cases/StatObjectQuery.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                }
+
+                StatObjectArgs args = new StatObjectArgs()
+                                                .WithBucket(bucketName)
+                                                .WithObject(bucketObject)
+                                                .WithVersionId(versionID)
+                                                .WithMatchETag(matchEtag)
+                                                .WithNotMatchETag(notMatchEtag)
+                                                .WithModifiedSince(modifiedSince)
+                                                .WithUnModifiedSince(unModifiedSince);
```
Comment (poornas) [inline Minio.Examples/Cases/StatObjectQuery.cs:66]:
```suggestion
                StatObjectArgs args = new StatObjectArgs()
                                                .WithBucket(bucketName)
                                                .WithObject(bucketObject)
                                                .WithVersionId(versionID)
                                                .WithMatchETag(matchEtag)
                                    //          .WithNotMatchETag(notMatchEtag)
                                                .WithModifiedSince(modifiedSince);
                                   //           .WithUnModifiedSince(unModifiedSince);
```

File: Minio.Examples/Cases/StatObjectQuery.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                                     string versionID = null,
+                                     string matchEtag = null,
+                                     string notMatchEtag = null,
+                                     DateTime modifiedSince = default(DateTime),
+                                     DateTime unModifiedSince = default(DateTime))
+        {
+            try
+            {
+                Console.WriteLine("Running example for API: StatObjectAsync [with ObjectQuery]");
+                if (!string.IsNullOrEmpty(matchEtag) && !string.IsNullOrEmpty(notMatchEtag))
```
Comment (poornas) [inline Minio.Examples/Cases/StatObjectQuery.cs:48]: remove these 2 if statements

**Review by poornas — CHANGES_REQUESTED** [review]

**Comment by poornas** [comment]
API.md will need an update

File: Minio.Examples/Cases/StatObjectQuery.cs (hunk 5 above)
Comment (BigUstad) [reply Minio.Examples/Cases/StatObjectQuery.cs:44]: Change made. But checks are added in `Validate()`

File: Minio.Examples/Cases/StatObjectQuery.cs (hunk 7 above)
Comment (BigUstad) [reply Minio.Examples/Cases/StatObjectQuery.cs]: Removed

**Review by poornas — CHANGES_REQUESTED** [review]

File: Minio/DataModel/ObjectOperationsResponse.cs [hunk 8]
```diff
@@ -30,4 +35,63 @@ internal SelectObjectContentResponse(HttpStatusCode statusCode, string responseC
...
+                else if (parameter.Name.Equals("ETag", StringComparison.OrdinalIgnoreCase))
+                {
+                    etag = parameter.Value.ToString().Replace("\"", string.Empty);
+                }
+                else if (parameter.Name.Equals("Content-Type", StringComparison.OrdinalIgnoreCase))
+                {
+                    contentType = parameter.Value.ToString();
+                    metaData["Content-Type"] = contentType;
+                }
+                else if (OperationsUtil.IsSupportedHeader(parameter.Name))
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/DataModel/ObjectOperationsResponse.cs]: Changed it.

File: Minio/DataModel/ObjectOperationsResponse.cs [hunk 9]
```diff
@@ -30,4 +35,63 @@ internal SelectObjectContentResponse(HttpStatusCode statusCode, string responseC
...
+                }
+                else if (parameter.Name.Equals("x-amz-version-id", StringComparison.OrdinalIgnoreCase))
+                {
+                    versionId = parameter.Value.ToString();
+                }
+                else if (parameter.Name.Equals("x-amz-delete-marker", StringComparison.OrdinalIgnoreCase))
+                {
+                    deleteMarker = parameter.Value.ToString().Equals("true");
+                }
+                else if (OperationsUtil.IsSupportedHeader(parameter.Name))
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/DataModel/ObjectOperationsResponse.cs]: Removed

File: Minio.Functional.Tests/FunctionalTest.cs [hunk 10]
```diff
@@ -1457,11 +1492,14 @@ internal async static Task CopyObject_Test8(MinioClient minio)
...
+                                            filestream, filestream.Length, "application/octet-stream", metaData:new Dictionary<string, string>{{"Orig", "orig-val with  spaces"}});
                 }
-                ObjectStat stats = await minio.StatObjectAsync(bucketName, objectName);
+                StatObjectArgs statObjectArgs = new StatObjectArgs()
+                                                        .WithBucket(bucketName)
+                                                        .WithObject(objectName);
+                ObjectStat stats = await minio.StatObjectAsync(statObjectArgs);
 
-                Assert.IsTrue(stats.MetaData["Orig"] != null) ;
+                Assert.IsTrue(stats.MetaData["orig"] != null) ;
```
Comment (poornas) [inline Minio.Functional.Tests/FunctionalTest.cs:1502]: why is this changing - the metadata key is case sensitive

File: Minio.Functional.Tests/FunctionalTest.cs [hunk 11]
```diff
@@ -1474,8 +1512,11 @@ internal async static Task CopyObject_Test8(MinioClient minio)
                 };
                 await minio.CopyObjectAsync(bucketName, objectName, destBucketName, destObjectName, copyConditions:copyCond, metadata: metadata);
 
-                ObjectStat dstats = await minio.StatObjectAsync(destBucketName, destObjectName);
-                Assert.IsTrue(dstats.MetaData["Mynewkey"] != null);
+                statObjectArgs = new StatObjectArgs()
+                                            .WithBucket(destBucketName)
+                                            .WithObject(destObjectName);
+                ObjectStat dstats = await minio.StatObjectAsync(statObjectArgs);
+                Assert.IsTrue(dstats.MetaData["mynewkey"] != null);
```
Comment (poornas) [inline Minio.Functional.Tests/FunctionalTest.cs:1519]: Again, here it is specifically checking that metadata key is case sensitive

**Review by poornas — CHANGES_REQUESTED** [review]

File: Minio.Examples/Cases/StatObject.cs [hunk 12]
```diff
@@ -15,29 +15,81 @@
...
+                    }
+                    return;
+                }
+                StatObjectArgs args = new StatObjectArgs()
+                                                .WithBucket(bucketName)
+                                                .WithObject(bucketObject)
+                                                .WithVersionId(versionID);
+                ObjectStat statObjectVersion = await minio.StatObjectAsync(args);
+                PrintStat(bucketObject, statObjectVersion);
+                if (showMetaData)
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio.Examples/Cases/StatObject.cs]: Removed flag. Showing metadata.

File: Minio/DataModel/ObjectOperationsArgs.cs [hunk 13]
```diff
@@ -86,4 +86,49 @@ public SelectObjectContentArgs WithRequestProgress(RequestProgress requestProgre
...
+        {
+            base.Validate();
+            if (!string.IsNullOrEmpty(this.NotMatchETag) && !string.IsNullOrEmpty(this.MatchETag))
+            {
+                throw new ArgumentException(nameof(this.NotMatchETag) + " and " + nameof(this.MatchETag) + " being set is not allowed.");
+            }
+            if ((this.ModifiedSince != null && !this.ModifiedSince.Equals(default(DateTime))) &&
+                    (this.ModifiedSince != null && !this.UnModifiedSince.Equals(default(DateTime))))
+            {
+                throw new ArgumentException(nameof(this.NotMatchETag) + " and " + nameof(this.MatchETag) + " being set is not allowed.");
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio/DataModel/ObjectOperationsArgs.cs]: Changed error message to reflect the appropriate message.

File: Minio.Functional.Tests/FunctionalTest.cs (hunk 10 above)
Comment (BigUstad) [reply Minio.Functional.Tests/FunctionalTest.cs]: Metadata dictionary with comparer IgnoreCase. Code updated.

File: Minio.Functional.Tests/FunctionalTest.cs (hunk 11 above)
Comment (BigUstad) [reply Minio.Functional.Tests/FunctionalTest.cs]: Metadata dictionary with comparer IgnoreCase. Code updated.

**Review by poornas — APPROVED** [review]
LGTM

**Review by kannappanr — APPROVED** [review]
LGTM

**Pull request merged**